rift/rib.py

```diff
@@ -19,56 +20,138 @@ def debug(self, msg, *args):
         if self._log:
             self._log.debug("[%s] %s" % (self._log_id, msg), *args)
 
-    def get_route(self, prefix, owner):
-        packet_common.assert_prefix_address_family(prefix, self.address_family)
-        if prefix in self.destinations:
-            return self.destinations[prefix].get_route(owner)
+    def get_route(self, rte_prefix, owner):
+        packet_common.assert_prefix_address_family(rte_prefix, self.address_family)
+        prefix = str(rte_prefix)
+        if self.destinations.has_key(prefix):
+            return self.destinations.get(prefix).get_route(owner)
         else:
             return None
 
     def put_route(self, rte):
+        """
+        Add a RibRoute object to the Destination object associated to the prefix.
+        :param rte: (RibRoute) the object to add to the Destination associated to the prefix
+        :return:
+        """
+
         packet_common.assert_prefix_address_family(rte.prefix, self.address_family)
         rte.stale = False
-        prefix = rte.prefix
         self.debug("Put %s", rte)
-        if rte.prefix in self.destinations:
-            destination = self.destinations[prefix]
+        # If there is no Destination object for the prefix, create a new Destination object
+        # for the given prefix and insert it in the Trie
+        prefix = packet_common.ip_prefix_str(rte.prefix)
+        if not self.destinations.has_key(prefix):
+            prefix_destination = _Destination(self, rte.prefix)
+            self.destinations.insert(prefix, prefix_destination)
         else:
-            destination = _Destination(prefix)
-            self.destinations[prefix] = destination
-        destination.put_route(rte, self.fib)
-
-    def del_route(self, prefix, owner):
-        # Returns True if the route was present in the table and False if not.
-        packet_common.assert_prefix_address_family(prefix, self.address_family)
-        if prefix in self.destinations:
-            destination = self.destinations[prefix]
-            deleted = destination.del_route(owner, self.fib)
-            if destination.routes == []:
-                del self.destinations[prefix]
+            prefix_destination = self.destinations.get(prefix)
+        # Insert desired route in destination object
+        best_changed = prefix_destination.put_route(rte)
+
+        # Get children prefixes before performing actions on the prefix
+        # (it can be deleted from the Trie)
+        children_prefixes = self.destinations.children(
+            packet_common.ip_prefix_str(prefix_destination.prefix))
+
+        # If best route changed in Destination object
+        if best_changed:
+            # Update prefix in the fib
+            self.fib.put_route(prefix_destination.best_route)
+            # Try to delete superfluous children
```

Comment:
Can you please explain what the concept of "deleting superfluous children" is?  I have not heard of the concept of a "superfluous" route before. Can you please explain what the definition of a "superfluous route" is, and why we need to delete them when the best route in the RIB changes.

It seems that this is intended to be special-case code for when the best route in the RIB changes. In my old code, I keep it simple (= low complexity = low chance of bugs): I simply delete the old route  in the FIB and install the new one.

Reply:
This method (`_delete_superfluous_children`) is triggered when a prefix becomes unreachable and all its children prefixes are removed from the RIB and the FIB. However, we made another commit that changes this method since we noticed that the children prefixes should be removed only from the FIB (but the RIB should keep them).

Reply:
Right, that's exactly what I expected - remove from FIB but not RIB